### PR TITLE
feat(lessons): add wildcard and regex pattern support for keyword matching

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Add support for wildcards (`*`) in keyword matching and a new `patterns` field for full regex matching in lesson frontmatter.

### Wildcards in Keywords

`*` in keywords matches zero or more word characters (`\w*`):

```yaml
match:
  keywords:
    - "process killed at * seconds"  # Matches "process killed at 120 seconds"
    - "timeout*"                      # Matches "timeout", "timeout30s", "timeouts"
```

### Patterns Field (Full Regex)

New `patterns` field for cases where full regex is needed:

```yaml
match:
  keywords:
    - "simple literal match"
    - "wildcard * match"
  patterns:
    - "error\\s+code\\s+\\d{3,4}"  # Full regex when needed
```

### Design Decisions (per discussion in gptme-contrib#102)

- `*` matches **empty strings** (zero or more chars) - `\w*` not `\w+`
- **No `?`** for single character (not needed initially)
- **Case insensitive** matching continues as current behavior
- **Separate `patterns` field** avoids YAML escaping confusion in keywords

### Files Changed

- `gptme/lessons/parser.py`: Add `patterns` field to `LessonMetadata`
- `gptme/lessons/matcher.py`: Add wildcard and pattern matching logic with caching
- `gptme/context/selector/rule_based.py`: Add wildcard and pattern support for consistency
- `tests/lessons/test_wildcard_matching.py`: Comprehensive tests (20 tests)

## Testing

All 20 tests pass:
- Unit tests for wildcard pattern conversion
- Unit tests for keyword matching with wildcards
- Unit tests for regex pattern matching
- Integration tests with `LessonMatcher` class

Closes gptme/gptme-contrib#102

----

> [!IMPORTANT]
> Adds wildcard and regex pattern support for keyword matching in lessons, updating matchers and parsers, with comprehensive tests.
> 
>   - **Behavior**:
>     - Adds wildcard (`*`) support in keywords, matching zero or more word characters (`\w*`).
>     - Introduces `patterns` field for full regex matching in lesson frontmatter.
>     - Wildcards do not match across spaces, hyphens, or punctuation.
>   - **Files and Functions**:
>     - `gptme/_keyword_matching.py`: Implements `_keyword_to_pattern`, `_compile_pattern`, `_match_keyword`, and `_match_pattern` for keyword and pattern matching.
>     - `gptme/lessons/parser.py`: Adds `patterns` field to `LessonMetadata` and updates `parse_lesson()` to handle new field.
>     - `gptme/lessons/matcher.py`: Updates `LessonMatcher` to use new matching functions for keywords and patterns.
>   - **Testing**:
>     - `tests/lessons/test_wildcard_matching.py`: Adds 20 tests for wildcard and pattern matching, including edge cases and parser validation.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=gptme%2Fgptme&utm_source=github&utm_medium=referral)<sup> for ea576d45cff59b646cef1a5ac26a869bc47fa62f. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
## Addressed Review Feedback

Pushed fix addressing Greptile review comments (d1486bd6f):

### Fixed Issues

1. **`match_keywords()` wildcard support** - Now uses `_match_keyword` for wildcard-aware matching
2. **`hybrid_matcher.py` consistency** - Updated `_keyword_score()` to use `_match_keyword`
3. **`matcher_enhanced.py` consistency** - Updated keyword matching to use `_match_keyword`
4. **Missing tests** - Added tests for:
   - Parsing `patterns` field from YAML frontmatter (3 tests)
   - `match_keywords` with wildcard support (3 tests)

Total tests now: **26** (up from 20)

### Consistency
All matcher files now consistently use `_match_keyword` for wildcard-aware keyword matching.

@greptileai review

**Comment by TimeToBuildBob** [comment]
## Addressed Code Review Feedback

Pushed commit ebe65e215 addressing Greptile feedback:

1. **Moved case normalization inside `_keyword_to_pattern`** - Cleaner API while preserving cache efficiency (normalized keys = fewer cache misses). Updated in both `matcher.py` and `rule_based.py`.

2. **Fixed confusing test comment** - The comment previously said "Should NOT match" then immediately explained why it DOES match. Clarified to accurately describe the expected behavior.

All 26 wildcard/pattern tests passing ✅

**Comment by TimeToBuildBob** [comment]
## Addressed Greptile Code Review Feedback

Pushed commit 1f61de0fd addressing remaining issues from Greptile's 3/5 confidence review:

### Changes Made

1. **Fixed Code Duplication** (main issue)
   - Extracted shared functions to `gptme/_keyword_matching.py`:
     - `_keyword_to_pattern()`
     - `_compile_pattern()`
     - `_match_keyword()`
     - `_match_pattern()`
   - Both `matcher.py` and `rule_based.py` now import from shared module
   - Eliminates ~80 lines of duplicated code

2. **Added Edge Case Validation**
   - Empty keywords/patterns return `None` (no match)
   - Single wildcard `*` correctly matches zero or more word characters
   - Whitespace-only strings treated as empty

3. **Improved Parser Validation**
   - Filter non-string patterns and keywords during YAML parsing
   - Support single string as alternative to list format
   - Skip empty/whitespace-only values

4. **Expanded Test Coverage**
   - Added 9 new tests for edge cases (35 total, up from 26):
     - `TestEdgeCases`: 5 tests for empty/whitespace/single-wildcard
     - `TestParserValidation`: 4 tests for non-string filtering

All 35 tests passing ✅

@greptileai review

**Comment by ErikBjare** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
## Response to Greptile Review

Thanks for the detailed analysis! I'd like to clarify that the issues raised have been addressed:

### ✅ Issues Already Resolved in Previous Commits

1. **`match_keywords()` wildcard support** - Already uses `_match_keyword()` helper (see lines 245-247 in `matcher.py`):
   ```python
   for lesson_kw in lesson.metadata.keywords:
       if _match_keyword(lesson_kw, input_kw):
           matched_keywords.append(input_kw)
   ```

2. **`matcher_enhanced.py`** - Already imports and uses `_match_keyword` from shared module (line 10, 106)

3. **`hybrid_matcher.py`** - Already imports and uses `_match_keyword` from shared module (line 7, 202)

4. **Test coverage for `patterns` field parsing** - Tests exist in `TestParsePatternsFromYAML` class (lines 227-279) and `TestParserValidation` class (lines 429-523)

### 🔧 Fixed in This Commit

- **Test imports** - Updated test file to import from the shared `gptme/_keyword_matching.py` module instead of `gptme.lessons.matcher`

### Test Results
All 35 tests pass:
```
============================== 35 passed in 0.80s ==============================
```

The 3/5 score may reflect analysis of an earlier commit. The current implementation addresses all concerns about wildcard support consistency across the codebase.

**Comment by TimeToBuildBob** [comment]
## Greptile Review Feedback Addressed (70fe52cd8)

Thanks for the thorough review! Made the following improvements:

### Cache Efficiency
- **Fixed**: Normalization now happens *before* the cache lookup
- `_keyword_to_pattern()` normalizes and delegates to `_keyword_to_pattern_cached()`
- Different cases ("Error", "error") now share the same cache entry

### Single Wildcard Safety
- **Fixed**: Single `*` now returns `None` to prevent over-matching all text
- Updated test to verify this safer behavior

### ReDoS Risk
- **Documented**: Added security note in module docstring explaining that patterns come from trusted sources (lesson files)
- This is an acceptable risk given the trust model (users control which lessons are loaded)

### Redundant .lower() Discussion
- The `.lower()` calls on `message_lower` and `query_lower` are actually needed for the `in` string checks (lines 112, 125, 75) that don't use regex
- The keyword/pattern lowercasing has been fixed by the cache efficiency improvement above

@greptileai review

**Comment by TimeToBuildBob** [comment]
## CI Failure Investigation

The CI failure on run `20880097824` is **not related to PR #1093 code changes**.

### Root Cause
The failure was a GitHub Actions runner infrastructure issue:
```
System.IO.IOException: No space left on device : '/home/runner/actions-runner/cached/_diag/Worker_20260110-150015-utc.log'
```

### Evidence
All actual test steps passed:
- ✓ Run tests
- ✓ Upload coverage reports to Codecov
- ✓ Upload test results to Codecov

The failure occurred during runner cleanup when the GitHub Actions runner ran out of disk space on its internal logs.

### Action Taken
Re-triggered CI with `gh run rerun --failed`. This should succeed with a fresh runner that has sufficient disk space.

---
This is a transient infrastructure issue, not a code bug.

**Comment by ErikBjare** [comment]
@TimeToBuildBob Use your new PR workflow lesson to address the remaining unresolved review comments.

File: gptme/_keyword_matching.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    # Normalize to lowercase for consistent caching
+    keyword = keyword.lower().strip()
+
+    if "*" in keyword:
+        # Escape special regex chars except *, then replace * with \w*
+        # First escape everything, then un-escape \* and replace with \w*
+        escaped = re.escape(keyword)
+        # re.escape converts * to \*, so we replace \* with \w*
+        pattern_str = escaped.replace(r"\*", r"\w*")
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/_keyword_matching.py:66]: ✅ Fixed in commit ea576d45c

Added explicit documentation about wildcard limitation:
- Note in `_keyword_to_pattern` docstring explaining `*` uses `\w*` which matches only word characters
- Clarified that `*` will NOT match across spaces, hyphens, or punctuation
- Updated example for single wildcard behavior

File: gptme/_keyword_matching.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        return re.compile(pattern, re.IGNORECASE)
+    except re.error as e:
+        logger.warning(f"Invalid regex pattern '{pattern}': {e}")
+        return None
+
+
+def _match_keyword(keyword: str, text: str) -> bool:
+    """Check if a keyword matches the text.
+
+    Uses word boundary matching to prevent partial word matches.
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/_keyword_matching.py]: ✅ Fixed in commit ea576d45c

Updated `_match_keyword` docstring to accurately describe the behavior:
- Changed from 'Uses word boundary matching' to 'Performs substring matching'
- Clarified that 'err' will match both 'error' and 'berry'
- Added note about wildcard character class limitation

File: gptme/_keyword_matching.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        keyword: Keyword to search for (may contain wildcards)
+        text: Text to search in
+
+    Returns:
+        True if keyword is found in text
+    """
+    pattern = _keyword_to_pattern(keyword)
+    if pattern is None:
+        return False
+    return pattern.search(text) is not None
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/_keyword_matching.py:124]: ✅ Fixed in commit ea576d45c

Updated docstring to match implementation (substring matching, not word boundaries). The current behavior is intentional - substring matching is more flexible for lesson keyword matching. If stricter word boundary matching is needed in the future, it could be added as an optional parameter.

**Comment by TimeToBuildBob** [comment]
## Addressed Remaining Review Comments

**Commit**: ea576d45c

Fixed the misleading documentation about word boundary matching:

| Comment | Issue | Resolution |
|---------|-------|------------|
| Line 66 (Greptile) | Wildcard `*` limitation may surprise users | ✅ Added explicit documentation |
| Line null (Ellipsis) | Docstring says 'word boundary' but no `\b` | ✅ Updated to 'substring matching' |
| Line 124 (Greptile) | Same - word boundary claim but no anchors | ✅ Fixed docstring to match implementation |

### Changes Made

1. **`_keyword_to_pattern` docstring**: Added explicit note about wildcard limitation:
   - `*` uses `\w*` (matches word characters only)
   - Will NOT match across spaces, hyphens, or punctuation
   - Updated single wildcard example

2. **`_match_keyword` docstring**: Corrected description:
   - Changed 'word boundary matching' to 'substring matching'
   - Clarified that 'err' matches both 'error' and 'berry'
   - Added note about wildcard character class

All 35 tests passing ✅

**Comment by ErikBjare** [comment]
Wouldn't it be better if `_keyword_to_pattern` matched across hyphens, punctuation, etc. Just not whitespace/newline or special delimiter-like chars such as `'"[](){}`? What would be most intuitive and useful while being precise?

Similarly about `_match_keyword` change to "substring matching", was this always the case? Not what I expected, do the same consideration there.

**Comment by TimeToBuildBob** [comment]
## Analysis of Matching Behavior

@ErikBjare - Good questions. Let me clarify:

### 1. Substring Matching - Was Always the Case

**Master branch** uses simple substring matching:
```python
for keyword in lesson.metadata.keywords:
    if keyword.lower() in message_lower:  # <-- substring matching
```

So yes, `err` matching `error` and `berry` is **not new** - it's how lessons always worked. The old docstring claiming 'word boundary matching' was simply **incorrect documentation**. My commit ea576d45c just corrected the docstring to match actual behavior.

### 2. Wildcard Pattern Improvement - Agree with Your Suggestion

**Current** (`\w*`): `*` matches only word characters (a-zA-Z0-9_)
- `error*message` would NOT match `error-message` or `error:message`

**Your suggestion** (match everything except whitespace/delimiters):
- `error*message` WOULD match `error-message`, `error:message`, `error.message`
- `error*message` would NOT match `error message` (space) or `error[message]` (delimiter)

This is more intuitive! Proposed implementation:
```python
# Instead of: r'\w*'
# Use: r'[^\s'"\[\](){}}]*'  # Anything except whitespace and delimiters
```

### Questions for You

1. **Substring matching**: Should I change this to word boundary matching, or keep the existing behavior? Word boundaries have their own issues (don't work well with underscores, punctuation).

2. **Wildcard improvement**: Want me to implement the change to match across punctuation/hyphens? This would be a behavior change but aligns with user intuition.

Let me know and I'll implement accordingly.

**Comment by ErikBjare** [comment]
I see now. Substring matching it is then. I was just concerned about how much a `*` would really match, could be a really long message and it's probably not meant to match across lines or so.

Is the Wildcard improvement really a good idea? Seems like a messy custom character group you made, if there's no good built-in regex special group (or combination of a couple, or a good best practice for situations like this) then that makes me hesitant.

**Comment by TimeToBuildBob** [comment]
## Wildcard Pattern - Keep Simple

@ErikBjare - You're right, there's no clean built-in regex group for "non-whitespace except delimiters."

**Options**:
1. `\w*` (current) - Word characters only. Simple, predictable, well-understood.
2. `\S*` - All non-whitespace. Too greedy (would match `[]` etc).
3. Custom `[^\s'"\[\](){}]*` - Works but ugly, hard to maintain.

**Recommendation**: Keep `\w*`. It's:
- Simple and well-documented
- Predictable behavior (users know `\w` from regex basics)
- Easy to explain: "wildcards match word characters"

The main limitation (not matching `error-message`) is minor - users can just use `error message` (substring match handles it) or `error` (shorter keyword).

If you agree, I'll:
1. Keep current `\w*` implementation
2. Ensure docs clearly explain the limitation
3. PR is ready for merge

Let me know!
